hudi-common/src/main/java/org/apache/hudi/metadata/AbstractHoodieTableMetadata.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hudi.metadata;
+
+import org.apache.hudi.common.config.SerializableConfiguration;
+import org.apache.hudi.common.engine.HoodieEngineContext;
+import org.apache.hudi.common.fs.FSUtils;
+import org.apache.hudi.common.table.HoodieTableConfig;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.util.PartitionPathEncodeUtils;
+import org.apache.hudi.common.util.StringUtils;
+import org.apache.hudi.exception.TableNotFoundException;
+import org.apache.hudi.expression.ArrayData;
+import org.apache.hudi.hadoop.CachingPath;
+import org.apache.hudi.hadoop.SerializablePath;
+import org.apache.hudi.internal.schema.Types;
+import org.apache.hudi.internal.schema.utils.Conversions;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public abstract class AbstractHoodieTableMetadata implements HoodieTableMetadata {
+
+  protected final transient HoodieEngineContext engineContext;
+
+  protected final SerializableConfiguration hadoopConf;
+  protected final SerializablePath dataBasePath;
+
+  protected final boolean hiveStylePartitioningEnabled;
+  protected final boolean urlEncodePartitioningEnabled;
+
+  // TODO get this from HoodieConfig
+  protected final boolean caseSensitive = false;
+
+  public AbstractHoodieTableMetadata(HoodieEngineContext engineContext, SerializableConfiguration conf, String dataBasePath) {
+    this.engineContext = engineContext;
+    this.hadoopConf = conf;
+    this.dataBasePath = new SerializablePath(new CachingPath(dataBasePath));
+    FileSystem fs = FSUtils.getFs(dataBasePath, conf.get());
+    Path metaPath = new Path(dataBasePath, HoodieTableMetaClient.METAFOLDER_NAME);
+    TableNotFoundException.checkTableValidity(fs, this.dataBasePath.get(), metaPath);
+    HoodieTableConfig tableConfig = new HoodieTableConfig(fs, metaPath.toString(), null, null);
```

Comment:
@bvaradar I found that `FileSystemBackedTableMetadata` is hotly used in many codes, so creating metaClient to get TableConfig could badly affect the performance(Making tests timeout).

So I change here to create TableConfig directly.

Reply:
@boneanxs : I think the slowness was due to loadActiveTimelineOnLoad=true when constructing HoodieTableMetaClient. If so, can you set it to False and use HoodieTableMetaClient. If possible, would prefer to go through HoodieTableMetaClient here to keep it similar with HoodieBackedTableMetadata

Reply:
I think it's not relate to `loadActiveTimelineOnLoad`, since it's default false, let me investigate why metaClient could be slow here.

Reply:
Is it possible that constructing `Configuration` could be the slowness? It needs to copy all properties  and synchronize the whole class `Configuration` (running into lock contention?).

```java
public Configuration(Configuration other) {
    this.resources = (ArrayList<Resource>) other.resources.clone();
    synchronized(other) {
      if (other.properties != null) {
        this.properties = (Properties)other.properties.clone();
      }

      if (other.overlay!=null) {
        this.overlay = (Properties)other.overlay.clone();
      }

      this.restrictSystemProps = other.restrictSystemProps;
      if (other.updatingResource != null) {
        this.updatingResource = new ConcurrentHashMap<String, String[]>(
           other.updatingResource);
      }
      this.finalParameters = Collections.newSetFromMap(
          new ConcurrentHashMap<String, Boolean>());
      this.finalParameters.addAll(other.finalParameters);
    }

    synchronized(Configuration.class) {
      REGISTRY.put(this, null);
    }
    this.classLoader = other.classLoader;
    this.loadDefaults = other.loadDefaults;
    setQuietMode(other.getQuietMode());
  }
```

`HoodieTableMetaClient` could call `new Configuration(configuration` 2 times, `this.hadoopConf = new SerializableConfiguration(conf);` and `getFs()`.

I do trace the construct method of `AbstractHoodieTableMetadata`:

1. Create HoodieTableMetaClient(1ms):

![Screenshot 2023-06-08 at 17 09 29](https://github.com/apache/hudi/assets/10115332/212b9c4e-9d8c-4452-ab17-9106e8b3bc91)

2. Create HoodieTableConfig(0.5ms including getFs and checkTableValidity):
![Screenshot 2023-06-08 at 17 07 37](https://github.com/apache/hudi/assets/10115332/e5275538-080b-4aa9-b120-7c5485ca323d)

and trace the construct method of `HoodieTableMetaClient`
![Screenshot 2023-06-08 at 20 44 51](https://github.com/apache/hudi/assets/10115332/f5feec9c-c544-4eeb-ada1-82552fc61593)

and the method `getFs`:

![Screenshot 2023-06-08 at 20 48 01](https://github.com/apache/hudi/assets/10115332/e81e2110-4400-4948-b85c-ac3c963c0e04)

we can see there're extra effort to initialize SerializedConfiguration(which is actually new Configuration)

Reply:
hmmm, We are not creating HoodieTableMetaClient so many times to cause this overhead to be the reason. right ?